File: gen_bspline_ndvi.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jul 23 11:25:16 2024

@author: Gabriel
"""

import pandas as pd
import numpy as np
from scipy.interpolate import BSpline
from patsy import dmatrix
import statsmodels.api as sm

pivoted_data  = pd.read_csv(r"C:\Projetos\bioflore\ecosia\i.txt")

label = 'ndvi_mean'

# Ensure dates are in datetime format
pivoted_data['date'] = pd.to_datetime(pivoted_data['date'])

# Sort the DataFrame by date
pivoted_data = pivoted_data.sort_values(by='date')

# Find the first non-NA date for the given label
date_non_na = pivoted_data.loc[~pivoted_data[label].isna(), 'date'].iloc[0]

# Define the specific range for the x-axis
start_date = pivoted_data.iloc[0]['date']
end_date = pivoted_data.iloc[-1]['date']

# Forward fill interpolation
pivoted_data[label] = pivoted_data[label].fillna(method='ffill')

# Prepare data for spline fitting
x = pivoted_data[label]
y = pivoted_data['date']

# Convert date to numeric for spline fitting
y_numeric = (y - y.min()).dt.days

# Degree and degrees of freedom
degree = 4
df = 15

# Fit B-spline
spline_basis = dmatrix("bs(train, degree=degree, df=df, include_intercept=False)", {"train": y_numeric}, return_type='dataframe')
fit_bs = sm.OLS(x, spline_basis).fit()

# Predictions for a smooth curve
y_pred = pd.date_range(start=date_non_na, end=end_date, freq='D')
y_pred_numeric = (y_pred - y.min()).days
spline_basis_pred = dmatrix("bs(train, degree=degree, df=df, include_intercept=False)", {"train": y_pred_numeric}, return_type='dataframe')
x_pred = fit_bs.predict(spline_basis_pred)

# Creating the resulting DataFrame
l = [label] * len(x_pred)
result_df = pd.DataFrame({'date': y_pred, 'ndvi_mean': x_pred, 'talhao': l})

import matplotlib.pyplot as plt
import seaborn as sns

# Ensure seaborn style for better aesthetics
sns.set()

# Plotting the result_df
plt.figure(figsize=(12, 6))

# Plot the original data
plt.plot(pivoted_data['date'], pivoted_data[label], 'o', label='Original data', markersize=5)

# Plot the smoothed curve
plt.plot(result_df['date'], result_df['ndvi_mean'], label='Smoothed B-spline fit', color='red', linewidth=2)

# Adding labels, title, legend, etc.
plt.xlabel('Date')
plt.ylabel('NDVI Mean')
plt.title('NDVI Mean Over Time with B-Spline Fit')
plt.legend()
plt.grid(True)

# Show plot
plt.show()

File: gen_bspline_ndvi.R
library(dplyr)
library(splines)
library(zoo)
library(tidyverse)
rm(list = ls())

tab<-read.table("C:/Projetos/bioflore/ecosia/resultado_22062024.csv", sep=",")
colnames(tab)<-tab[1,]
tab<-tab[2:nrow(tab),]
tab$ndvi_mean<-as.numeric(tab$ndvi_mean)

labels <- as.list(unique(tab$talhao))

result <- data.frame(date=as.Date(character()),
                 ndvi_mean=numeric(), 
                 talhao=character(), 
                 stringsAsFactors=FALSE)

for (label in labels){
  #label = "ESP_RE"
  print(label)
  vc <- c(label)
  data<-tab[tab$talhao %in% vc,]
  data <- data[!duplicated(data[c('talhao','date')]),]
  
  pivoted_data <- data %>% pivot_wider(
    names_from = talhao,
    values_from = ndvi_mean
  )
  # Prepare data for plotting
  pivoted_data$date <- as.Date(pivoted_data$date) # Ensure dates are in Date format
  pivoted_data <- pivoted_data[order(pivoted_data$date),]
  
  date_non_na <- pivoted_data %>% filter(!is.na(pivoted_data[[label]])) %>% slice(1) %>% pull(date)
  #pivoted_data %>% filter(date >= date_non_na)
  # Define the specific range for the x-axis
  start_date <- pivoted_data[1,1]$date 
  end_date <- pivoted_data[nrow(pivoted_data),1]$date
  
  # Open a new graphics window (specific to Windows)
  
  x <- pivoted_data[[label]]
  # Forward fill interpolation
  x <- na.locf(x, na.rm = FALSE)
  y <- pivoted_data$date
  #y_re <- IPE_RE$date
  degree <- 15 # Degree of the spline
  df <- 5 # Degrees of freedom or roughly the number of knots
  
  # Fit B-spline
  fit_bs <- lm(x ~ bs(y, degree = degree, df = df))
  # Predictions for a smooth curve across the entire date range
  y_pred <- seq(date_non_na, end_date, by = "day")
  x_pred <- predict(fit_bs, newdata = list(y = y_pred))
  l = rep( label , length( x_pred)  ) 
  
  df <- data.frame(y_pred,
                   x_pred,
                   l)
  names(df) <- c('date','ndvi_mean','talhao')
  #print(df)
  result <- rbind(result, df)
}

write.csv(result, "C:\\Projetos\\bioflore\\ecosia\\Ecosia\\bspline_ndvi.csv", row.names=FALSE)
#combined_df <-  do.call(rbind, df_list)
#dev.off()
